### PR TITLE
Immediate fallback for MultiDatacentenderMessageProducer

### PR DESCRIPTION
**Pull request by moscicky**
Until now fallback to remote DC was executed after `speculativeSendDelay` elapsed. This meant that all fallbacks were executed after `speculativeSendDelay` ms even if the local send failed after <1ms. In worst case scenario we were waiting up to `speculativeSendDelay` waiting for the fallback to execute.

 In this PR `immediate` fallback is introduced which performs fallback as soon as request to local DC fails and cancels the execution of scheduled (speculative) fallback if possible

### CODE REVIEW

**Review by piotrrzysko — APPROVED** [review]

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/config/ConsumerSenderConfiguration.java [hunk 1]
```diff
@@ -265,7 +265,8 @@ public FutureAsyncTimeout futureAsyncTimeoutFactory(InstrumentedExecutorServiceF
         ScheduledExecutorService timeoutExecutorService = executorFactory.getScheduledExecutorService(
                 "async-timeout",
                 senderAsyncTimeoutProperties.getThreadPoolSize(),
-                senderAsyncTimeoutProperties.isThreadPoolMonitoringEnabled()
+                senderAsyncTimeoutProperties.isThreadPoolMonitoringEnabled(),
+                false
```
Comment (szczygiel-m) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/config/ConsumerSenderConfiguration.java:269]: What do You think about splitting `InstrumentedExecutorServiceFactory.getScheduledExecutorService()` to two methods (overloading it), one with boolean parameter `removeOnCancel`. Thanks to this we would not have misleading hardcoding parameters and changes in consumers.

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/config/FrontendProducerConfiguration.java [hunk 2]
```diff
@@ -111,6 +106,17 @@ public KafkaMessageSenders failFastKafkaMessageSenders(FailFastKafkaProducerProp
...
     }
 
+    @Bean
+    public ScheduledExecutorService chaosScheduler(KafkaChaosProperties chaosProperties, InstrumentedExecutorServiceFactory executorServiceFactory) {
+        KafkaChaosProperties.ChaosSchedulerProperties chaosSchedulerProperties = chaosProperties.getChaosScheduler();
+        return executorServiceFactory.getScheduledExecutorService(
+                "chaos",
+                chaosSchedulerProperties.getThreadPoolSize(),
+                chaosSchedulerProperties.isThreadPoolMonitoringEnabled(),
+                false
```
Comment (szczygiel-m) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/config/FrontendProducerConfiguration.java:116]: shouldn't this property be configurable in `chaosSchedulerProperties`?

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/producer/kafka/MultiDatacenterMessageProducer.java [hunk 3]
```diff
@@ -60,86 +59,115 @@ public void send(Message message, CachedTopic cachedTopic, PublishingCallback ca
...
+    We first try to send message to local DC. If the local send fails we perform 'immediate' fallback to remote DC.
+
+    Additionally, we schedule a 'speculative' fallback to remote DC to execute after 'speculativeSendDelay' elapses.
+    Speculative fallback decreases publication latency but may result in messages being duplicated across DCs.
 
-        sendOrScheduleChaosExperiment(
-                localSender,
+    If local DC send succeeds or fails before 'speculativeSendDelay' elapses we try to cancel the 'speculative' fallback if
+    it has not been executed yet. We guarantee that only one fallback executes - either 'immediate' or 'speculative'.
+     */
```
Comment (szczygiel-m) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/producer/kafka/MultiDatacenterMessageProducer.java:119]: thanks for detailed description 🥇

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/config/ConsumerSenderConfiguration.java (hunk 1 above)
Comment (moscicky) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/config/ConsumerSenderConfiguration.java:269]: Sure, I have created a builder which will be more future-proof, please take a look

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/executor/InstrumentedExecutorServiceFactory.java [hunk 4]
```diff
@@ -33,24 +33,51 @@ public ExecutorService getExecutorService(String name, int size, boolean monitor
...
 
-    public ScheduledExecutorService getScheduledExecutorService(
-            String name, int size, boolean monitoringEnabled, boolean removeOnCancel
-    ) {
 
-        ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat(name + "-scheduled-executor-%d").build();
-        ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(size, threadFactory);
-        executor.setRemoveOnCancelPolicy(removeOnCancel);
-        return monitoringEnabled ? monitor(name, executor) : executor;
+    public class ScheduledExecutorServiceBuilder {
```
Comment (szczygiel-m) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/executor/InstrumentedExecutorServiceFactory.java:37]: I was thinking about overloaded constructor, but builder seems to be even better

**Review by szczygiel-m — APPROVED** [review]

File: hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/config/FailFastLocalKafkaProducerProperties.java [hunk 5]
```diff
@@ -4,6 +4,44 @@
...
+    maximum frontend request duration (frontend.handlers.maxPublishRequestDuration). This means that when
+    maxPublishRequestDuration is exceeded for a message we received, a client will receive 5xx even tough the
+    corresponding message is still being processed in the producer. The message will eventually be ACKd by Kafka so upon client-side
+    retry the message will be duplicated. This however, would likely also happen if the message was promptly timed-out by producer
+    before maxPublishRequestDuration elapsed - the message was likely already sent to Kafka, there just haven't been a response yet.
+
+    So by using large requestTimeout we cause the first slow message to be duplicated (by client-side retry) but:
+    - we protect other inflight messages from being duplicated,
+    - we prevent connections from being frequently dropped and reestablished.
+ */
```
Comment (piotrrzysko) [inline hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/config/FailFastLocalKafkaProducerProperties.java:44]: Great comment! Very helpful.

**Review by piotrrzysko — DISMISSED** [review]

**Review by szczygiel-m — APPROVED** [review]

**Review by piotrrzysko — APPROVED** [review]

**Pull request merged**